### PR TITLE
Add gradient checkpointing for Song Unet

### PR DESCRIPTION
**Pull request by daviddpruitt**
# Modulus Pull Request

## Description
Adds the option for gradient checkpointing in the SongUnet

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: examples/generative/corrdiff/conf/config_train_base.yaml [hunk 1]
```diff
@@ -54,6 +54,8 @@ bench: false
   # Enable cuDNN benchmarking
 workers: 4
   # DataLoader worker processes
+songunet_checkpoint_level: 0
+  # Gradient checkpointing level, value is number of layers to checkpoint
```
Comment (mnabian) [inline examples/generative/corrdiff/conf/config_train_base.yaml:58]: Can we clarify that 0 means no checkpointing?

File: test/models/diffusion/test_song_unet.py [hunk 2]
```diff
@@ -63,6 +63,26 @@ def test_song_unet_forward(device):
...
+        in_channels=2,
+        out_channels=2,
+        embedding_type="fourier",
+        channel_mult_noise=2,
+        encoder_type="residual",
+        resample_filter=[1, 3, 3, 1],
+        checkpoint_level=4,
+    ).to(device)
+
+    assert common.validate_forward_accuracy(
```
Comment (mnabian) [inline test/models/diffusion/test_song_unet.py:78]: Would be great to also add a backward accuracy test. See this as an example: https://github.com/NVIDIA/modulus/blob/main/test/models/graphcast/test_grad_checkpointing.py#L25

File: examples/generative/corrdiff/conf/config_train_base.yaml (hunk 1 above)
Comment (daviddpruitt) [reply examples/generative/corrdiff/conf/config_train_base.yaml:58]: Updated with a note that indicates 0 is no checkpointing.

File: test/models/diffusion/test_song_unet.py (hunk 2 above)
Comment (daviddpruitt) [reply test/models/diffusion/test_song_unet.py]: Rewrote the gradient checkpointing tests as a seperate suite, now includes forward and backward.

**Comment by mnabian** [comment]
@daviddpruitt just checking if this is ready for review?

**Review by mnabian — APPROVED** [review]

**Review by akshaysubr — APPROVED** [review]

**Pull request merged**